File: ATM/python/atm.py
class user_account:
    def __init__(self,name,account_number,phone_number,intial_amount):
        self.name = name
        self.account_number = account_number
        self.phone_number = phone_number
        self.balance =  intial_amount
    def withdraw(self,amount):
        if(amount<=self.balance):
            self.balance -= amount
            return 1
        else:
            return 0
    def deposit(self,amount):
        self.balance += amount
    def check_balance(self):
        return self.balance

#Maintains a global dictionary of the the account holders corresponding to their user_account object
Accounts = {}

#maintains a global dictionary of the account_number corresponding to their password
accountno_password = {}

#maintains the account number of the currently logged in user
global_account_number = ""

def create_account(username,phone_number,password,intial_amount,account_number):
    new_user = user_account(username, account_number, phone_number, intial_amount)
    #Storing the new_user object into dictionary using account_no as a key 
    Accounts[account_number] = new_user
    #storing the password into dictionary using account_no as a password
    accountno_password[account_number] = password
    return account_number
    
def login(account_number,password):
    #verfying the user account_number againt the password using accountno_password dictionary
    if(accountno_password[account_number]==password):
        global_account_number = account_number
        return True

def logout():
    #making the global_account_number to null indicates no user has been logged in
    global_account_number = ""

def change_password(account_number,password):
    #verifying account_number against the password
    if(accountno_password[account_number]==password):
        new_password = input("enter new password: ")
        accountno_password[account_number] = new_password
        print("********************************************************\n")
        print("password changed successfully\n")
        print("********************************************************\n")
    else:
        print("Invalid password")

#for filling sample data
def fill_data():
    name = ["vamsi","saketh","bindu","uday"]
    phone_numbers = ["123","124","12345","2345"]
    account_numbers = ["1","2","3","4"]
    passwords = ["a","b","c","d"]
    intial_amounts = [120,150,100,140]
    for i in range(4):
        create_account(name[i], phone_numbers[i], passwords[i], intial_amounts[i], account_numbers[i])

    
fill_data()
while(True):
    print("********************************************************\n")
    print("Welcome to GVPAI Bank\n")
    print("********************************************************\n")
    print("1.Login\n2.Exit")
    choice = int(input())
    if(choice==1):
        account_number = input("Enter Account number: ")
        password = input("Enter password: ")
        #checking whether the entered account_number is valid or not
        if(account_number in Accounts.keys()):
            #checking whether the entered password is valid or not
            if(login(account_number, password)):
                global_account_number = account_number
                logged_in_user_obj = Accounts[global_account_number]
                print("********************************************************\n")
                print("welcome back "+logged_in_user_obj.name+"\n")
                print("********************************************************\n")
                innerchoice = True
                while(innerchoice):
                    print("1.check balance\n2.Deposit\n3.withdraw\n4.change password\n5.Logout\n")
                    print("********************************************************\n")
                    user_choice = int(input("Enter choice: "))
                    if(user_choice==1):
                        balance = logged_in_user_obj.check_balance()
                        print("********************************************************\n")
                        print("Balance = ",balance,"\n")
                        print("********************************************************\n")
                    elif(user_choice==2):
                        deposit_amount = int(input("Enter deposit amount: "))
                        logged_in_user_obj.deposit(deposit_amount)
                        print("********************************************************\n")
                        print("Balance updated\n")
                        print("********************************************************\n")
                    elif(user_choice == 3):
                        withdraw_amount = int(input("Enter withdraw amount: "))
                        status = logged_in_user_obj.withdraw(withdraw_amount)
                        if(status==1):
                            print("********************************************************\n")
                            print("Balance updated\n")
                            print("********************************************************\n")
                        else:
                            print("********************************************************\n")
                            print("cannot perform this operation\n")
                            print("********************************************************\n")
                    elif(user_choice==4):
                        password = input("Enter current password: ")
                        change_password(global_account_number, password)
                    elif(user_choice==5):
                        logout()
                        break
                    else:
                        print("********************************************************\n")
                        print("Invalid Choice\n")
                        print("********************************************************\n")
            else:
                print("********************************************************\n")
                print("Invalid password\n")
                print("********************************************************\n")
        else:
            print("********************************************************\n")
            print("Invalid Account number\n")
            print("********************************************************\n")
    else:
        break
        
    


    

File: ATM/java/Atm.java
import java.util.*;
public class Atm {
    static Scanner sc = new Scanner(System.in);
    static HashMap<String,User_account> Accounts = new HashMap<>();
    static HashMap<String,String> acc_pass = new HashMap<>();
    static String global_user_acc_no = "";

    public static void create_account(String name,String phone_number,String account_number,String password,int intial_amount){
        User_account new_user = new User_account(name,phone_number,account_number,intial_amount);
        Accounts.put(account_number,new_user);
        acc_pass.put(account_number,password);
    }

    public static int login(String account_number,String password){
        if(acc_pass.get(account_number).equals(password)){
            global_user_acc_no = account_number;
            return 1;
        }
        return 0;
    }

    public static void logout(){
        global_user_acc_no = "";
    }

    public static void change_password(String account_number,String password){
        if(acc_pass.get(account_number).equals(password)){
            System.out.println("Enter new password");
            String new_password = sc.next();
            acc_pass.put(account_number,new_password);
            System.out.println("*******************************\n");
            System.out.println("Password succesfully changed\n");
            System.out.println("*******************************\n");
        }
        else{
            System.out.println("*******************************\n");
            System.out.println("Invalid details\n");
            System.out.println("*******************************\n");
        }
    }

    public static void fill_data(){
        ArrayList<String> names = new ArrayList<String>(Arrays.asList("vamsi", "saketh", "bindu","uday"));
        ArrayList<String> phone_numbers = new ArrayList<String>(Arrays.asList("123", "124", "12345","2345"));
        ArrayList<String> account_numbers = new ArrayList<String>(Arrays.asList("1", "2", "3","4"));
        ArrayList<String> passwords = new ArrayList<String>(Arrays.asList("a", "b", "c","d"));
        ArrayList<Integer> inital_amounts = new ArrayList<Integer>(Arrays.asList(120,150,100,140));
        for(int i=0;i<4;i++){
            create_account(names.get(i),phone_numbers.get(i),account_numbers.get(i),passwords.get(i),inital_amounts.get(i));
        }
    }

    public static void main(String s []){
        System.out.println("*******************************\n");
        System.out.println("Welcome to GVPAI Bank\n");
        System.out.println("*******************************\n");
        fill_data();
        boolean flag = true;
        while(flag){
            System.out.println("1.Login\n2.Exit");
            System.out.println("Enter choice");
            int choice = sc.nextInt();
            switch (choice){
                case 1: System.out.println("Enter Account number: ");
                        String account_number= sc.next();
                        System.out.println("Enter password: ");
                        String password = sc.next();
                        if(Accounts.containsKey(account_number)) {
                            if (login(account_number, password) == 1) {
                                User_account current_user = Accounts.get(account_number);
                                System.out.println("*******************************\n");
                                System.out.println("Welcome back "+current_user.name+"\n");
                                System.out.println("*******************************\n");
                                boolean innerflag = true;
                                while(innerflag) {
                                    System.out.println("1.Check Balance\n2.Deposit\n3.Withdrawal\n4.changepassword\n5.Logout");
                                    System.out.println("Enter choice: ");
                                    int innerchoice = sc.nextInt();
                                    switch (innerchoice) {
                                        case 1:
                                            int balance = current_user.check_balance();
                                            System.out.println("*******************************\n");
                                            System.out.println("Balance = " + balance + "\n");
                                            System.out.println("*******************************\n");
                                            break;
                                        case 2:
                                            System.out.println("Enter Deposit amount : ");
                                            int deposit_amount = sc.nextInt();
                                            current_user.deposit(deposit_amount);
                                            System.out.println("*******************************\n");
                                            System.out.println("Balance updated\n");
                                            System.out.println("*******************************\n");
                                            break;
                                        case 3:
                                            System.out.println("Enter withdraw amount : ");
                                            int withdraw_amount = sc.nextInt();
                                            int status = current_user.withdraw(withdraw_amount);
                                            if (status == 1) {
                                                System.out.println("*******************************\n");
                                                System.out.println("Balance updated\n");
                                                System.out.println("*******************************\n");
                                            }
                                            else{
                                                System.out.println("*******************************\n");
                                                System.out.println("Insufficient Balance\n");
                                                System.out.println("*******************************\n");
                                            }
                                            break;
                                        case 4:
                                            System.out.println("Enter current password : ");
                                            String curr_password = sc.next();
                                            change_password(global_user_acc_no, curr_password);
                                            break;
                                        case 5:
                                            logout();
                                            innerflag = false;
                                            break;
                                        default:
                                            System.out.println("*******************************\n");
                                            System.out.println("Invalid choice\n");
                                            System.out.println("*******************************\n");
                                    }
                                }
                            } else {
                                System.out.println("*******************************\n");
                                System.out.println("Invalid password\n");
                                System.out.println("*******************************\n");
                            }
                        }
                        else{
                            System.out.println("*******************************\n");
                            System.out.println("Invalid Account number\n");
                            System.out.println("*******************************\n");
                        }
                        break;
                case 2: flag = false;
                        break;
                default:
                    System.out.println("*******************************\n");
                    System.out.println("Invalid choice\n");
                    System.out.println("*******************************\n");
            }
        }
    }
}


